README.md
[Home](https://qb64.com) • [Forums](https://qb64.boards.net/) • [News](news.md) • [GitHub](https://github.com/QB64Official/qb64) • [Wiki](wiki.md) • [Samples](samples.md) • [InForm](inform.md) • [GX](gx.md) • [QBjs](qbjs.md) • [Community](community.md) • [More...](more.md)

<img src="images/qb64.png" alt="TUI Interface" title="TUI Interface" style="display:block;margin-left:auto;margin-right:auto">

**Our [forums have](https://qb64.boards.net/) returned, check them out [here](https://qb64.boards.net/)!**

## Get QB64

Please select the correct package for your OS.

### Version 2.1

Please download the correct package for your operating system.

| Operating System | Package |
| - | - |
| Windows x64 | [Download](https://github.com/QB64Official/qb64/releases/download/v2.1/qb64_dev_2022-09-08-07-14-00_47f5044_win-x64.7z) |
| Windows x86/32-bit | *See below.* |
| Linux | [Download](https://github.com/QB64Official/qb64/releases/download/v2.1/qb64_dev_2022-09-08-07-14-00_47f5044_lnx.tar.gz) |
| macOS | [Download](https://github.com/QB64Official/qb64/releases/download/v2.1/qb64_dev_2022-09-08-07-14-00_47f5044_osx.tar.gz) |

*Windows x86/32-bit: Due to an error during the build process, the 32-bit build of QB64 v2.1 is not available; however, this issue is confirmed to be fixed with the next release (v2.1.1). We apologize for the inconvenience this may have caused and ask for your understanding as we continue to iron out the finer details related to the overall release process.*

> NOTE: For additional instructions on how to setup/configure, please visit the [source repo](github.md).

## Introduction

**How BASIC made its way into the 21st century**

<img src="images/colorwin10.png" alt="TUI Interface" title="Settings" style="float:right;width:214px;height:144px;padding:6px">

The BASIC language has been the gateway into programming for countless people. Popular as a beginner programming language in the 80’s and evolving into a powerful professional tool in 90’s, BASIC (and its successor QBasic), helped many people develop a love for programming. These languages provided the foundational learning platform for most of today’s professional developers.

<img src="images/dllwin10.png" alt="TUI Interface" title="In this sample code, internal keywords are colored blue, metacommands and user procedures in green and strings in orange. Notice the DECLARE LIBRARY block used to access Windows' API." style="float:left;width:214px;height:144px;padding:6px">

The QB64 project has evolved over the last decade to bring the magic and educational potential of BASIC from its 20th century roots into the modern era. The QB64 project is already in use in both educational and professional contexts and has an active and helpful user community.

Unlike traditional BASIC and QBasic code, QB64 gets compiled automatically into machine code – allowing exceptional performance, easy distribution, and the ability to link with external C and C++ programming libaries. Compatible with most QBasic 4.5 code, QB64 adds a number of extensions, such as OpenGL and other modern features, providing the perfect blend of classic and modern program development.

QB64 is available for all recent Windows, Linux, and macOS versions.

## Who created QB64?

For some of the early history of QB64, check out [these interviews](galleon.md).

## Additional/Complementary Tools

There are several people in the community with projects that serve to compliment QB64; extending the reach and capability of the QB64 developer.

- [InForm](inform.md): Rapid Application Development (GUI) for QB64.
- [GX](gx.md): A game engine for QB64 that also allows you to extend your reach to the web.
- [QBjs](qbjs.md): An implementation of the BASIC programming language for the web, with multimedia support and easy sharing of programs that aims compatibility with QBasic, QB4.5 and QB64.
- [QB64 Interpreter](https://github.com/FellippeHeitor/QB64-interpreter): Run QB64 commands on the fly or load a file and run it, no compilation required - written in QB64.
- [L-BASIC](https://github.com/flukiluke/L-BASIC): The L-BASIC compiler and interpreter implemented in QB64.  There is also this [forum thread](https://qb64forum.alephc.xyz/index.php?topic=2778.0) for some interesting background on the project.
- [Roslyn](https://github.com/dotnet/roslyn): Open-source MIT-licensed implementation of latest direct *"commercial"* decendant of QBasic/QB4.5 that targets the [.NET](https://dotnet.microsoft.com/) platform.
- [PC-BASIC](https://robhagemans.github.io/pcbasic/): Free, cross-platform emulator for the GW-BASIC family of interpreters.

### What is InForm?

![InForm1](images/inform_designer_v1_3.png)

[InForm](inform.md) is a Rapid Application Development tool for QB64. It consists of a library of graphical routines and a WYSIWYG editor that allows you to design forms and export the resulting code to generate an event-driven QB64 program.

Want to build graphical UI applications across Windows, Linux and/or Mac? Check out [InForm](inform.md)!

### What is GX?

[GX](gx.md) is a basic game engine... literally. This is a Game(G) Engine(X) built with and for QB64, a QBasic/QuickBASIC IDE and compiler with modern extensions. [GX](gx.md) supports basic 2D gaming: platformer, top-down, etc.; you know, classic NES/SNES type games.

Interested in extending your QB64 reach to the web?  Check out [GX](gx.md).

### What is QBjs?

[![QBjs](images/qbjs.png)](https://qbjs.org)

[QBjs](qbjs.md) can be considered a sort of *sister* project of QB64. It is heavily inspired by folks that have a huge appreciation for QB64 and the main developer that is working on this is "simply" continuing forward on a project that originally grew out of a pet project of his for QB64. This (previous) project, written in QB64, allows you to write your code in QB64 and then "convert" it to Javascript. The [QBjs](qbjs.md) project is taking this further by providing the necessary tools to write your code in the browser directly.

Interested in playing with QBasic/QB64 in the browser?  Check out [QBjs](qbjs.md).

### What is QB64 Interpreter?

Run QB64 commands on the fly or load a file and run it, no compilation required. Written in QB64!  Check it out at [QB64 Interpreter](https://github.com/FellippeHeitor/QB64-interpreter).

### What is L-BASIC?

[L-BASIC](https://github.com/flukiluke/L-BASIC) is a new BASIC language and compiler; a language variant that is reasonably close to QB64. However, it doesn't attempt to be 100% compatibile; willing to break compatibility with programs from 1985 when needed.  What makes this project interesting is that not only is it written in QB64, but it serves to explore ideas that "could be". This isn't to suggest that what [L-BASIC](https://github.com/flukiluke/L-BASIC) is doing will make it's way into QB64, but it does serve as an interesting thought experiment and visible example of what you could do if you desired to start with QB64 and "make it your own".  Check it out at [L-BASIC](https://github.com/flukiluke/L-BASIC).

### What is Roslyn?

Although a bit controversial to include this on this list, [Roslyn](https://github.com/dotnet/roslyn) notably includes self-hosting versions of the Visual Basic for [.NET](https://dotnet.microsoft.com/) compiler – a compiler written in the language itself. The compiler is available via the traditional command-line programs but also as APIs available natively from within [.NET](https://dotnet.microsoft.com/) code. [Roslyn](https://github.com/dotnet/roslyn) exposes modules for syntactic (lexical) analysis of code, semantic analysis, dynamic compilation to CIL, and code emission. This project was started in 2010, made open source in 2014 (Apache License 2.0) and released first version in Visual Studio 2015. At some point the license was transitioned to use the same license as [.NET](https://dotnet.microsoft.com/) (MIT-licensed). You can either work with this directly from command-line tools via [.NET](https://dotnet.microsoft.com/) or by installing the *Community Edition* (Free) of [Visual Studio](https://visualstudio.microsoft.com/). It is included in this list as a lot of the skills / foundation gained by working in QB64 can easily applied to VB and the scope of what can be done with [.NET](https://dotnet.microsoft.com/) is pretty impressive.

### What is PC-BASIC?

[PC-BASIC](https://robhagemans.github.io/pcbasic/): Free, cross-platform emulator for the GW-BASIC family of interpreters that allows you to run classic games and legacy BASIC applications designed for MS-DOS systems, IBM PC, PCjr or Tandy 1000. [PC-BASIC](https://robhagemans.github.io/pcbasic/) aims for "bug-for-bug compatibility with Microsoft GW-BASIC". In other words, it has a very similar mission to QB64, but instead of QBasic and/or QB4.5 compatibility [PC-BASIC](https://robhagemans.github.io/pcbasic/) is focused on the BASIC that was prior to QBasic/QB4.5 ([GW-BASIC](https://gw-basic.com/)) and is an excellent alternative/replacement to [GW-BASIC](https://gw-basic.com/) on modern platforms. Because of this, it might be able to execute *line-number* BASIC programs that aren't compatible with QBasic, QB4.5 and QB64.

## FAQ

### What about the qb64.net website/domain?

Please make note that the .net domain for QB64 is highly suspect. Apparently the story goes that someone forgot to renew the domain name and it was snipped by a questionable party that subsequently mirrored the previous websites information and spammed it with dubious advertising. At this point it has nothing to do with QB64 and, ultimately, should be completely avoided given that the content could contain suspect binaries potentially containing malware. Additionally, it's been stated in the community has potentially having malware as part of the site - potentially infecting your machine by simply browsing to it. So, in the end, if you do venture there be sure to exercise caution.


news.md
[Home](https://qb64.com) • [Forums](https://qb64.boards.net/) • [News](news.md) • [GitHub](https://github.com/QB64Official/qb64) • [Wiki](wiki.md) • [Samples](samples.md) • [InForm](inform.md) • [GX](gx.md) • [QBjs](qbjs.md) • [Community](community.md) • [More...](more.md)

### 2022-09-09

*BETA* Release 0.7.0 of [QB64 Extension for Visual Studio Code](https://github.com/QB64Official/vscode/tree/development/releases) available. This thing just keeps getting better and better!

### 2022-09-08

**QB64 releases v2.1!** See details [here](https://github.com/QB64Official/qb64/releases/tag/v2.1). Note that this release is marked as *Pre-release* due to the significant changes related to the build/release process.

### 2022-09-07

Congratulations to @dbox for capturing **SILVER** (second) in the [Jam For All Basic Dialects #3](https://itch.io/jam/jam-for-all-basic-dialects-3/results) representing QB64 and his GX game engine!

### 2022-09-05

*BETA* [QB64 for NP++ v0.2.0](https://github.com/QB64Official/notepadplusplus/releases) released!

### 2022-09-04

*BETA* Things are becoming more polished and functional with the release 0.6.2 of [QB64 Extension for Visual Studio Code](https://github.com/QB64Official/vscode/tree/development/releases). (Note: This is only available in the development branch.)

### 2022-08-26

*BETA* Although still very much a beta, the 0.5.0 build of [QB64 Extension for Visual Studio Code](https://github.com/QB64Official/vscode/tree/main/releases) is really starting to take shape.

### 2022-08-15

*BETA* Use Notepad++, then you should really check out [QB64 for NP++ v0.1.0](https://github.com/QB64Official/notepadplusplus/releases).

### 2022-08-11

*BETA* The first, very early, release of [QB64 Extension for Visual Studio Code](https://github.com/QB64Official/vscode/tree/main/releases) is now available through a manual install (not available in the VSCode Marketplace at this time).

### 2022-06-09

The [Jam For All Basic Dialects #3](https://itch.io/jam/jam-for-all-basic-dialects-3) contest has been announced!

> It is easier to learn than many other programming languages as BASIC's commands are similar to English and it has a simple set of rules. Nowadays there are many BASIC dialects which are suitable for game development. Many of them are free of charge. I list some BASIC dialects which are suitable for game development here: (But you can use any other BASIC dialects which are not on the following list for this game jam.)

> The jam will be suitable for any kind of game developer and you will have so much fun in this jam.  Furthermore,  all entrants can share their experiences of using BASIC dialects in the jam. Before the jam starts, three themes will be selected from the following six themes by me.(1. War  2. Random  3. Escape  4. Powerful  5. One Way Trip  6. Circles, Squares & Triangles)

The contest will take during the month of August, so be sure to check it out!

See Also: [Jam For All Basic Dialects #1](https://itch.io/jam/jam-for-all-basic-dialects), [Jam For All Basic Dialects #2](https://itch.io/jam/jam-for-all-basic-dialects-2)

-- Cory Smith

### 2022-06-04

Some planning has taken place and we can announce that work has begun on first fixing the issues surrounding the content behind the wiki and how this will be made available, maintained and ensure stability. Work has begun but it will take some time complete the full transition; with that said, the effort should worth it in the end as it helps to better future proof the content.

-- Cory Smith

### 2022-07-06

[QBjs](qbjs.md) has just [released v0.5.0-beta](https://github.com/boxgaming/qbjs/discussions/32) that adds `Data`, `Read` and `Restore`, a significant parser update, improved error reporting and 35 new keywords!

### 2022-05-18

[QBjs](qbjs.md) has just [released v0.4.0-beta](https://github.com/boxgaming/qbjs/discussions/24) that adds simple sharing via url embedding, standard libraries and a host of additions, fixes and updates.

You can see some of this work in action by checking out these samples:


samples/lfx/src/lfx.bas
''''''''''''''''''''''''''''''
'LFX by Jon Mark O'Connor    '
'                            '
''''''''''''''''''''''''''''''
'Copyright (C) 1995 DOS World
'Published in Issue #20, March 1995


DECLARE SUB LAYER1 ()        'You don't need
DECLARE SUB LAYER2 ()        'to type in these
DECLARE SUB LETTERS ()       'DECLARE SUBs. They're
DECLARE SUB NORMAL1 ()       'added when you save
DECLARE SUB NORMAL0 ()       'the program.
DECLARE SUB TRANSLATE ()     '

DefInt A-Z 'The program starts here.
Dim P$(122)
Common Shared P$(), P$, SIZE$
Common Shared C1$, C2$, C3$
Common Shared C4$, BGC, DOT$
Common Shared SPACING$, WORD$
Common Shared PFT, PFL
Call LETTERS 'Call letterset.
Screen 9 '''''''''''''''''''''Set screen mode to 9.

EFFECT1:
Line (1, 1)-(338, 72), 4, BF 'Effect background.
Line (140, 3)-(336, 70), 1, BF
WORD$ = "DOS World" 'Word(s) to be converted.
SPACING$ = "L2" 'L decreases spacing; R increases it.
SIZE$ = "24" 'Use increments of four, up to half of resolution.
DOT$ = "N" 'No dot over i and j.
BGC = 4 'Background color.
Call TRANSLATE 'Converts above to a DRAW command.
C1$ = "C14": C2$ = "C 8": C3$ = "C 0": C4$ = "C15" 'Colors.
PFT = 52: PFL = 13: Call LAYER1 'PFT = position from top.
PFT = 52: PFL = 9: Call LAYER1 'PFL = position from left.
PFT = 51: PFL = 10: Call LAYER1 'Increase and decrease
PFT = 51: PFL = 11: Call LAYER1 'PFT and PFL to create depth.

EFFECT2:
Line (3, 78)-(338, 130), 13, B
Line (5, 80)-(336, 128), 7, BF
WORD$ = "DOS World"
SPACING$ = "L1"
SIZE$ = "20"
DOT$ = "N"
BGC = 7
Call TRANSLATE
C1$ = "C 5": C2$ = "C 8": C3$ = "C 8": C4$ = "C 8"
PFT = 117: PFL = 17: Call LAYER1
PFT = 117: PFL = 15: Call LAYER1
PFT = 117: PFL = 13: Call LAYER1
C1$ = "C 5": C2$ = "C 0": C3$ = "C 0": C4$ = "C15"
PFT = 115: PFL = 19: Call LAYER1
PFT = 115: PFL = 18: Call LAYER1

EFFECT3:
Line (3, 136)-(214, 192), 8, BF
Line (3, 136)-(338, 192), 7, B
WORD$ = "embossed glass"
SPACING$ = "L2"
SIZE$ = "16"
DOT$ = "N"
BGC = 8
Call TRANSLATE
C1$ = "C 0": C2$ = "C 0": C3$ = "C 0": C4$ = "C 0"
PFT = 167: PFL = 12: Call LAYER1
C1$ = "C 8": C2$ = "C 7": C3$ = "C15": C4$ = "C 0"
PFT = 167: PFL = 13: Call LAYER2
PFT = 167: PFL = 12: Call LAYER2
PFT = 166: PFL = 13: Call LAYER2
PFT = 167: PFL = 11: Call LAYER2

EFFECT4:
Line (2, 206)-(636, 303), 12, BF
Line (1, 204)-(639, 306), 15, B
Line (19, 280)-(620, 280), 14, B
Line (20, 281)-(619, 281), 0, B
Line (21, 282)-(618, 282), 0, B
WORD$ = "Sugar's Shack"
SPACING$ = "L1"
SIZE$ = "28"
DOT$ = "N"
BGC = 12
Call TRANSLATE
C1$ = "C12": C2$ = "C 4": C3$ = "C 0": C4$ = "C 0"
PFT = 266: PFL = 24: Call LAYER1
C1$ = "C12": C2$ = "C 4": C3$ = "C 0": C4$ = "C 0"
PFT = 265
For XX = 25 To 33 Step 2
    PFL = XX: Call LAYER1
    PFT = PFT - 1
Next
C1$ = "C15": C2$ = "C 6": C3$ = "C 6": C4$ = "C14"
PFT = 259: PFL = 40: Call LAYER1
PFT = 259: PFL = 36: Call LAYER1
PFT = 258: PFL = 38: Call LAYER1

EFFECT5:
Line (402, 4)-(639, 176), 15, B
Line (403, 5)-(638, 175), 11, B
Line (404, 6)-(637, 174), 9, BF
WORD$ = "Lfx"
SPACING$ = "L2"
SIZE$ = "48"
DOT$ = "N"
BGC = 9
Call TRANSLATE
C1$ = "C 0": C2$ = "C 0"
PFT = 123
For XX = 408 To 438 Step 2
    PFL = XX: Call NORMAL1
PFT = PFT - 1: Next XX
PFT = PFT - 1
C1$ = "C15": C2$ = "C 8": C3$ = "C 7": C4$ = "C15"
For XX = 436 To 413 Step -1
    PFL = XX: Call LAYER1
PFT = PFT + 1: Next XX

AUTHOR:
WORD$ = "by Jon Mark O'Connor"
SPACING$ = "R1"
SIZE$ = " 4"
DOT$ = "N"
BGC = 9
Call TRANSLATE
C1$ = "C15": C2$ = "C 0"
PFT = 167: PFL = 440: Call NORMAL1

ALPHABET:
Line (1, 315)-(639, 331), 7, B
Line (2, 316)-(638, 330), 15, BF
WORD$ = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*()_-+=:;'."
SPACING$ = "R0"
SIZE$ = " 4"
DOT$ = "N"
BGC = 15
Call TRANSLATE
C1$ = "C 1"
PFT = 325: PFL = 5: Call NORMAL0

HITENTER:
Line (2, 337)-(78, 350), 2, BF
WORD$ = "Hit Enter"
SPACING$ = "R1"
SIZE$ = " 4"
DOT$ = "N"
BGC = 2
Call TRANSLATE
C1$ = "C14"
PFT = 346: PFL = 5: Call NORMAL1

INKEY1:
C1$ = "C14"
PFT = 346: PFL = 5: Call NORMAL1
Sleep 1
C1$ = "C10"
PFT = 346: PFL = 5: Call NORMAL1
Sleep 1
A$ = InKey$: If A$ = "" Then GoTo INKEY1
End

'''''''''''''''''''''' SUBS '''''''''''''''''''

Sub LAYER1
    PSet (PFL + 2, PFT + 5), BGC: Draw C3$ + P$
    PSet (PFL + 0, PFT + 5), BGC: Draw C3$ + P$
    PSet (PFL + 1, PFT + 5), BGC: Draw C3$ + P$
    PSet (PFL + 0, PFT + 4), BGC: Draw C3$ + P$
    PSet (PFL + 1, PFT + 2), BGC: Draw C2$ + P$
    PSet (PFL + 2, PFT + 4), BGC: Draw C1$ + P$
    PSet (PFL + 3, PFT + 3), BGC: Draw C4$ + P$
    PSet (PFL + 3, PFT + 2), BGC: Draw C2$ + P$
    PSet (PFL + 4, PFT + 4), BGC: Draw C2$ + P$
    PSet (PFL + 5, PFT + 2), BGC: Draw C3$ + P$
    PSet (PFL + 5, PFT + 3), BGC: Draw C2$ + P$
    PSet (PFL + 4, PFT + 3), BGC: Draw C1$ + P$
End Sub

Sub LAYER2
    PSet (PFL + 0, PFT + 5), BGC: Draw C4$ + P$
    PSet (PFL + 0, PFT + 3), BGC: Draw C4$ + P$
    PSet (PFL + 2, PFT + 3), BGC: Draw C1$ + P$
    PSet (PFL + 1, PFT + 2), BGC: Draw C1$ + P$
    PSet (PFL + 2, PFT + 4), BGC: Draw C4$ + P$
    PSet (PFL + 3, PFT + 3), BGC: Draw C3$ + P$
    PSet (PFL + 3, PFT + 2), BGC: Draw C1$ + P$
    PSet (PFL + 4, PFT + 4), BGC: Draw C1$ + P$
    PSet (PFL + 5, PFT + 2), BGC: Draw C1$ + P$
    PSet (PFL + 5, PFT + 3), BGC: Draw C2$ + P$
    PSet (PFL + 4, PFT + 3), BGC: Draw C1$ + P$
End Sub

Sub LETTERS
    P$(32) = "BR4" 'space
    P$(33) = "BR1BD1U1BU2U5BD7BR3" '!
    P$(34) = "BR0BU7D2BR2U2BD7BR3" '"
    P$(35) = "BD1BR1U8D2L1R5L1U2D8U2L4R5BR2BD1" '#
    P$(36) = "BD1BR3U8D8BL3BU5U1E1R3F1BD1BL5F1R3F1D1G1L3H1BR7BD1" '$
    P$(37) = "BD1U1E6U1BL5BD2U1R1D1L1BD4BR4R1D1L1U1BD1BR5" '%
    P$(38) = "BR1BU2BL1E3U1H1L1G1D1F6BU1BL5BU3G2D1F1R3E3BD2BR2" '&
    P$(39) = "BR2BU7D1G1BD5BR4" ''
    P$(40) = "BR1BD1BU2U4E2G2D4F2BU1BR4" '(
    P$(41) = "BR2BD1BU2U4H2F2D4G2BU1BR5" ')
    P$(42) = "BR2BU2U6D3L2R4BL4BD2E4BL4F4BD3BR4" '*
    P$(43) = "BR2BD1U7D3L2R5BD3BR3" '+
    P$(44) = "BD1U1R1D2G1BU3BR4" ',
    P$(45) = "BR1BU3R3L3BD3BR5" '-
    P$(46) = "BD1BR1L1BU1BR4" '.
    P$(47) = "BR1E5G5BR7" '/
    P$(48) = "BD1BU1U6E1R2F1D6G1L2H1BR6" '0
    P$(49) = "BD1BR2U7BL1E1D1BD6BR3" '1
    P$(50) = "BR1BU6E1R2F1D2G4D1R4BU1BR3" '2
    P$(51) = "BU6E1R2F1D2G1L2R2F1D2G1L2H1BR7" '3
    P$(52) = "BU3E4D8U4L4R5BD3BR3" '4
    P$(53) = "BR1BU3U4R4L4D4E1R2F1D3G1L2H1BR7" '5
    P$(54) = "BR1BD1BU4U3E1R2F1H1L2G1D6F1R2E1U3H1L2G1BD3BR6" '6
    P$(55) = "BR1BU7R4D3G4D1BU1BR7" '7
    P$(56) = "BR1BU6E1R2F1D2G1L2R2F1D2G1L2H1U2E1H1U2BD6BR7" '8
    P$(57) = "BR1BU4U2E1R2F1D6G1L2H1BU4F1R3BD3BR3" '9
    P$(58) = "BR1U1BU2U1BD4BR3" ':
    P$(59) = "BD1BR1BU1U1BU2U1BD4G1BU1BR6" ';
    P$(60) = "BU3E3G3F3BR3" '<
    P$(61) = "BU3R4BD2L4BR7BD1" '=
    P$(62) = "BU6BR1F3G3BR5" '>
    P$(63) = "BR1BU5U1E1R2F1D2G1D1BD2D1BU1BR3" '?
    P$(64) = "BD1BU1U6E1R3F1D5L2H1U2E1R1D3BD3L3H1F1R3E1BR3" '@
    P$(65) = "BD1U7E1R3F1D3L5R5D4BR4BU1" 'A
    P$(66) = "BD1U8R4F1D2G1L4R4F1D2G1L4BU1BR9" 'B
    P$(67) = "BD1BU1U6E1R3F1BD6G1L3H1F1R3BR4BU1" 'C
    P$(68) = "BD1U8R4F1D6G1L4BR9BU1 " 'D
    P$(69) = "BD1U8R4L4D4R3L3D4R4BU1BR4" 'E
    P$(70) = "BD1U8R5L5D4R4L4D4BR8BU1" 'F
    P$(71) = "BD1BU1U6E1R3F1BD6G1L3H1F1R3E1U3L2R2D3BR4" 'G
    P$(72) = "BD1U8D4R4U4D8BR4BU1" 'H
    P$(73) = "BD1BR3U8L1R2L1D8L1R2BR4BU1" 'I
    P$(74) = "BD1BU1F1R2E1U7L1R2L1D7BR5" 'J
    P$(75) = "BD1BU8D8U4R1E4G4F4BR4BU1" 'K
    P$(76) = "BD1BU8D8R4BU1BR4" 'L
    P$(77) = "BD1U7E1R2F1D7U7E1R2F1D7BU1BR4" 'M
    P$(78) = "BD1U8F7D1U8D8BR4BU1" 'N
    P$(79) = "BD1BU1U6E1R3F1BD6G1L3H1BR5U6D6BR4" 'O
    P$(80) = "BD1U8R4F1D2G1L4D4BR9BU1" 'P
    P$(81) = "BD1BU1U6E1R3F1BD6G1L3H1BR5U6BL3BD5F3BG3BR7BU5" 'Q
    P$(82) = "BD1U8R4F1D2G1L4R1F4BR4BU1" 'R
    P$(83) = "BD1BU5U2E1R3F1BD2BL5F1R3F1D2G1L3H1BR9" 'S
    P$(84) = "BR3BD1BU8L4R6L3D8BR6BU1" 'T
    P$(85) = "BD1BU1U7D7F1R3E1U7D7BR4" 'U
    P$(86) = "BD1BU8D5F2D1U1E2U5D5BR5BD2" 'V
    P$(87) = "BD1BU1U7D7F1R2E1U7D7F1R2E1U7D7BD1BR4BU1" 'W
    P$(88) = "BD1BU8D1F3E3U1D1G6D1U1E3F3D1BU1BR5" 'X
    P$(89) = "BL1BU5U2D2F3E3U2D2G3D3BU1BR7" 'Y
    P$(90) = "BU7R5D2G5D1R5BR4BU1" 'Z
    P$(91) = "BR1BD1BR3L3U8R3BR3BD7" '[
    P$(92) = "BU5F5BR3" '\
    P$(93) = "BD1R3U8L3BR2BD7BR4" ']
    P$(94) = "BR1BU5E2F2BD5BR2" '^
    P$(95) = "BD1R4BU1BL2BR5" '_
    P$(96) = "BR2BU7D1F1BD5BR2" '`
    P$(97) = "U2E1R3U1H1L2G1E1R2F1D4G1L2H1F1BR2E1U2D3BR4BU1" 'a
    P$(98) = "BD1U8D8R3E1U4H1L2G1D4BR8" 'b
    P$(99) = "BR0U4E1R2F1BD4G1L2H1F1R2E1BR4" 'c
    P$(100) = "U4E1R2F1H1L2G1D4F1R2E1U7D7D1BU1BR4" 'd
    P$(101) = "U4E1R2F1D2L4D2F1R2E1BR4" 'e
    P$(102) = "BR1BD1U7E1R2F1H1L2G1D3L1R3BD3BR5" 'f
    P$(103) = "U4E1R2F1D4G1L2H1BD3F1R2E1U3BR4" 'g
    P$(104) = "BU7D8U5E1R2F1D5BU1BR4" 'h
    P$(105) = "BR3BD1U6BU2BD7BR4" 'i
    P$(106) = "BR1BD1U6BU2BD7D3G1L2BU4BR8" 'j
    P$(107) = "BD1BU8D8U3E3G3F3BU1BR3" 'k
    P$(108) = "BR1BD1U8L1R1D8L1R2BR4BU1" 'l
    P$(109) = "BD1U6D1E1R1F1D5U5E1R1F1D5BU1BR4" 'm
    P$(110) = "BD1U6D1E1R2F1D5U4BD3BR4" 'n
    P$(111) = "U4E1R2F1D4G1L2H1BR8" 'o
    P$(112) = "U4E1R2F1D4G1L2H1D4U4BR8" 'p
    P$(113) = "U4E1R2F1D4G1L2H1BR4D4L1R2L1U4BR5" 'q
    P$(114) = "BD1U6D1E1R2F1H1L2G1U1D6BU1BR7" 'r
    P$(115) = "BU3U1E1R2F1H1L2G1D1F1R2F1D1G1L2H1BR8" 's
    P$(116) = "BR1BU7D2L2R4L2D5F1R1E1BR4" 't
    P$(117) = "U5D5F1R2E1U5D5BR4" 'u
    P$(118) = "BD1BU6D3F2D1U1E2U3D3BD2BR4" 'v
    P$(119) = "BD1BU6D5F1R1E1U4D4F1R1E1U5D5BR4" 'w
    P$(120) = "BD1BU6D1F2E2U1D1G4D1U1E2F2D1BU1BR4" 'x
    P$(121) = "BU5D5F1R2E1U5D8G1L2H1BU3BR8" 'y
    P$(122) = "BU5R4D1G4D1R4BU1BR4" 'z
End Sub

Sub NORMAL0
    PSet (PFL + 0, PFT + 0), BGC: Draw C1$ + P$
End Sub

Sub NORMAL1
    PSet (PFL + 0, PFT + 1), BGC: Draw C2$ + P$
    PSet (PFL + 1, PFT + 0), BGC: Draw C1$ + P$
    PSet (PFL + 2, PFT + 0), BGC: Draw C1$ + P$
End Sub

Sub TRANSLATE
    SPACING$ = "B" + UCase$(SPACING$)
    SIZE$ = "S" + UCase$(SIZE$)
    P$(105) = "BR0BD1U6BU2BD7BR4" 'i
    P$(106) = "BR1BD1U6BU2BD7D3G1L2BU4BR8" 'j
    If DOT$ = "Y" Then P$(105) = "BR0BD1U6BU1U1BD7BR4" 'Dot over i.
    If DOT$ = "Y" Then P$(106) = "BR1BD1U6BU1U1BD7D3G1L2BU4BR8" 'Dot over j.
    P$ = SIZE$
    For J = 1 To Len(WORD$)
        P$ = P$ + P$(Asc(Mid$(WORD$, J, 1))) + SPACING$
    Next J
End Sub

